---
title: "Class 13: Genome Informatics (Pt. 1)"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Ensemble 1000 Genome data

What proportion of the Mexican Ancestry in Los Angeles sample population (MXL)
are homozygous for the asthma associated SNP (G|G)?

```{r}
SNP <- read.csv("373531-SampleGenotypes-Homo_sapiens_Variation_Sample_rs8067378.csv", header = TRUE)
table(SNP$Genotype..forward.strand.)
#proportion
table(SNP$Genotype..forward.strand.)/nrow(SNP)
```


## Quality scores in FASTQ filses

```{r}
library(seqinr)
library(gtools)
phred <- asc( s2c("DDDDCDEDCDDDDBBDDDCC@") ) - 33
phred
```


## Population Scale Analysis

Read this file into R and determine the sample size for each genotype and their
corresponding median expression levels for each of these genotypes.

```{r}
snp_data <- read.table("rs8067378_ENSG00000172057.6.txt")
summary(snp_data)
```

```{r}
inds <- (snp_data$geno == "G/G")
snp_gg <- snp_data[inds,]
summary(snp_gg$exp)
snp_ag <- snp_data[(snp_data$geno == "A/G"),]
summary(snp_ag$exp)
snp_aa <- snp_data[(snp_data$geno == "A/A"),]
summary(snp_aa$exp)
```

First try at a boxplot to show this data

```{r}
boxplot(formula = snp_data$exp ~ snp_data$geno, notch = TRUE)
```
